Vsd - merge values

jezhiggins opened the pull request:
As a precursor to work on widening, I've been filling out the merge implementation for the various value representations. 

The changes are relatively minor, generally simplifying and generalising the merge. Once change of note is the addition of `internal_hash` and `internal_equality` members function to `interval_abstract_valuet`, which means intervals can be inserted into value-sets correctly, without duplication. 

The bulk of this PR is unit tests, covering the various combinations of constants, intervals, and value-sets.

jezhiggins committed:
constant_abstract_value merge test tidy-up

jezhiggins committed:
Tidying work on abstact_objectt and constant_abstract_valuet merge tests

Combine related scenarios. Use consistent language in THEN clauses. Use
consistent expectations.

jezhiggins committed:
Extend constant_abstract_valuet to merge with other value representations

Can successfully merge with matching intervals and value sets, including
from bottom. Goes TOP otherwise.

jezhiggins committed:
Add abstract_value_objectt::to_interval()

Analogous to to_constant, it returns a constant_interval_exprt that
expresses the range of the value. For constants the interval is across
whatever the constant is, for intervals it is their interval(!) and for
value_sets it is the interval between the smallest and largest values in
the set.

Used to intervals to merge with any other value type.

This commit includes unit tests for interval_abstract_objectt::merge.

jezhiggins committed:
Correct value_set merge implementation.

Ensure bottom flag is cleared when returning a new object. Intervals and
constants can be merged into the a set.

Extend abstract_object_set to_interval to handle intervals in the set

jezhiggins committed:
Add interval internal_hash/equality

Intervals can be added to value sets correctly now we're hashing on
value rather than identity.

jezhiggins commented:
The check-ubuntu-20_04-make-gcc build failure appear to be unrelated to the changes on this PR.

martin-cs commented on src/analyses/variable-sensitivity/value_set_abstract_object.cpp:
```diff
@@ -162,6 +163,18 @@ value_set_abstract_objectt::value_range_implementation() const
+{
+  verify();
+  return values.size() == 1 ? (*values.begin())->to_constant()
+                            : abstract_objectt::to_constant();
```
In the case of multiple entries, merge all of them then `to_constant`?

jezhiggins replied:
Add new test case merging 1 with { 1, [1, 1] } returns 1.

martin-cs commented on src/analyses/variable-sensitivity/abstract_object_set.cpp:
```diff
@@ -37,7 +37,7 @@ void abstract_object_sett::output(
 
-abstract_object_pointert abstract_object_sett::to_interval()
+constant_interval_exprt abstract_object_sett::to_interval() const
 {
```
Rather than using `to_constant` this should probably call `to_interval` so that if it is a value set of intervals it winds up taking the join of all of the intervals.

martin-cs replied:
Ah; ignore this, I have found the later commit that does this!

martin-cs commented on unit/analyses/variable-sensitivity/value_expression_evaluation/expression_evaluation.cpp:
```diff
@@ -595,9 +595,9 @@ SCENARIO(
       auto result = add_as_value_set(op1, op2, environment, ns);
 
-      THEN("= { [2,4], [2,3], [2,3], 2 }")
-      { // duplicate interval ok as first pass. Will be eliminated in due course.
```
Nice!

martin-cs commented on unit/analyses/variable-sensitivity/constant_abstract_value/#merge.cpp#:
```diff
@@ -0,0 +1,468 @@
+/*******************************************************************\
+
+ Module: Unit tests for constant_abstract_valuet::merge
```
File name? `#merge.cpp#`?  Maybe this shouldn't be in the PR?

jezhiggins replied:
Oops, sorry.

martin-cs commented on unit/analyses/variable-sensitivity/constant_abstract_value/#merge.cpp#:
```diff
@@ -0,0 +1,468 @@
+
+      THEN("result is unmodified BOTTOM")
+      {
+<<<<<<< HEAD
```
Yeah... don't think this is supposed to be here.

martin-cs commented on unit/analyses/variable-sensitivity/interval_abstract_value/merge.cpp:
```diff
@@ -453,6 +453,32 @@ SCENARIO(
+        EXPECT_MODIFIED(merged, val3, val6);
+      }
+    }
+    WHEN("merging [3, 4] with { 1, [3, 3] }")
```
Nice!

jezhiggins replied:
We aim to please.

martin-cs reviewed (changes requested):
One critical fix, the rest should be good to go.  If the unit test linking issue persists we should investigate.

martin-cs reviewed (approved):
Thanks for the fixes.

jezhiggins committed:
Improve value_set to_constant()

For multi-valued sets, calculate the interval then, if it's a single
value interval return a constant, otherwise use the base operation.
